dn1/src/dn1.jl

```diff
@@ -0,0 +1,155 @@
+module dn1
+
+export RedkaMatrika, sor
+
+import Base: getindex, setindex!, size, *, -
+
+"""
+    RedkaMatrika(V, I)
+
+podatkovni tip za redke matrike. Polja sta matriki velikosti nxm z neničelnimi elementi, kjer je m<=n ter matrika ki jo opisuje velikosti nxn. 
+V mattriki V so zapisane vrednosti neničelnih elementov v posameznih vrsticah. V matriki I so zapisani stolcpi istoležnih vvrednosti v posameznih vrsticah
+"""
+mutable struct RedkaMatrika
+    V
+    I
+end
+
+"""
+    n, n = size(R)
+
+funkcija vrne dimenzije redke matrike R
+"""
+function size(R::RedkaMatrika)
+    return (size(R.V)[1], size(R.V)[1])
+end
+
+"""
+    negR = -R
+
+funkcija implementira množenje redke matrike z -1
+"""
+function -(R::RedkaMatrika)
+    R.V = -R.V
+    return R
+end
+
+
+"""
+    a = R[i, i]
+
+pridobivanje vrednosti iz redke matrike R na določenem mestu
+"""
+function getindex(R::RedkaMatrika, i::Int, j::Int)
+    n, _ = size(R)
+    if (i > n || j > n || i < 1 || j < 1)
+        throw(BoundsError(R, (i, j)))
+    end
+    
+
+    if (j in R.I[i, :])
+        ix = findfirst(==(j), R.I[i, :])
+        return R.V[i, ix]
+    end
+
+    return 0
+end
+
+"""
+    R[i, j] = a
+
+funkcija za nastavljanje vrednosti redke matrike R na določenem mestu
+"""
+function setindex!(R::RedkaMatrika, v, i::Int, j::Int)
+    n, m = size(R.I)
+    if (i > n || j > n || i < 1 || j < 1)
+        throw(BoundsError(R, (i, j)))
+    end
+
+    if R[i, j] != 0
+        ix = findfirst(==(j), R.I[i, :])
+        R.V[i, ix] = v
+    else 
+
+        dummy_index = findfirst(==(nothing), R.I[i, :])
+        if dummy_index != nothing
+            R.I[i, dummy_index] = j
+            R.V[i, dummy_index] = v
+        else
+            R.I = hcat(R.I, fill(nothing, n))
+            R.V = hcat(R.V, fill(nothing, n))
+
+            R.I[i, end] = j
+            R.V[i, end] = v
+        end
+    end
+end
+
+"""
+    x = R*v
+
+funkcija za množenje redke matrike R z vektorjem v z desne
+"""
+function *(R::RedkaMatrika, v::Vector)
+    n, m = size(R.I)
+    result = zeros(eltype(v), n)
+    for i in 1:n #rows
+        for j in 1:m #entries in columns
+            col = R.I[i, j]
+            if col == nothing
+                continue
+            end
+            val = R.V[i, j]
+            result[i] += val * v[col]
+            
+        end
+    end
+
+    return result
+end
+
+"""
+    x_new = korak_gs(A, x, b)
+
+funkcija izračuna en korak Gaus-Seidlove metode ter vrne nov približek
+"""
+function korak_gs(A, x, b)
+    x_new = copy(x)
+    for i in 1:length(x)
+        sum_to = 0
+        sum_after = 0
+        for j in 1:i-1
+            sum_to += A[i, j] * x_new[j]
+        end
+        for j in i+1:length(x)
+            sum_after += A[i, j] * x[j]
```

Comment:
Namesto `x[j]` lahko uporabite kar `x_new[j]`, saj imata v tem trenutku isto vrednost. Tako lahko obe zanki združite v eno od `1:m`
```jl
sum = 0
for j in 1:m
   k = A.I[j]
   if k == nothing || k == i
     next
   end
   sum += A[i, k] * v[k]
 end
```

Lahko pa ustvarite pomožno funkcijo `nenicelni(A::RedkaMatrika, i)`, ki vrne indekse neničelnih elementov v i-ti vrstici in uporabite to

```jl
function nenicelni(A::RedkaMatrika, i)
   ...
   for j in 1:m
     if A.I[j] != nothing
       push!(indeksi, A.I[j])
     end
  end
  indeksi
end

function sor...
  ...
  for j in nenicelni(A, i)
    ...
```